### PR TITLE
Add cache req document

### PR DESCRIPTION
**Pull request by justin-stephenson**
I wrote this based on my understanding of reading the code, please let me know if I missed any important components and I will add it. For example I did not give a detailed explanation about domain selection/location codepaths.

### CODE REVIEW

**Comment by pbrezina** [comment]
I think this is a good start but imho the focus is different then what it should be.

The target reader is a developer (external and internal) who wants to contribute to the project and use cache_req in his code or even implement a cache_req plugin. So I think you don't need to focus on the NSS responder at all, but put all energy into describing the cache_req in detail - not only the basic search but all the details that happens under the hood, all the plugins capabilities and switches and so on.

The goal is to really describe how it works under the hood - even I don't understand it completely since it changed a lot from the time I wrote it :-) And ideally, make it easily maintenable.

BTW I described the basic lookup in SSSD architecture, you can just link it from this document when its merged and focus on details here.

**Comment by pbrezina** [comment]
I think this document should consist from two parts:

1. How to use cache_req API
2. What happens under the hood and how to write new plugins and extend functionality

**Comment by justin-stephenson** [comment]
Thank you, I'll re-work the document and submit an update once done.

**Comment by pbrezina** [comment]
> > I think this document should consist from two parts:
> > ```
> > 1. How to use cache_req API
> > ```
> 
> From internally within SSSD is acceptable? Or is this intended to be an external API for any application to call?

Internal only.

> > What happens under the hood and how to write new plugins and extend functionality
> 
> I thought that a cache_req plugin exists to handle each type of NSS library call, what would be the use case for writing a new plugin? Is `CACHE_REQ_USER_BY_CERT` an example of this type of use case?

It's not only about nss, it handles everything (expect sudo for now). So if something new is added, it needs to get support in cache req.

**Comment by pbrezina** [comment]
Perhaps, we can split it into two documents:
* how to use cache req API into "fundamentals" section
* how does it work internally into "under the hood" section

File: src/contrib/cache_req.rst [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+The ``cache req`` (request) interface provides an internal API to request identity
+information from SSSD, either directly from the local SSSD cache or retrieving refreshed data
+from the data provider. The SSSD NSS responder is the primary consumer of the interface,
+however ``cache_req`` also refreshes information which may be needed by the PAM
+responder, or provide user information attributes over the SSSD infopipe.
+
+.. mermaid::
+
+	graph LR
```
Comment (elkoniu) [inline src/contrib/cache_req.rst:15]: This chart looks "too flat" to me. I think similar to the one [1] made by jhrozek in the past would be more meaningful.
[1] https://jhrozek.files.wordpress.com/2015/03/sssd-lookup.png

File: src/contrib/cache_req.rst [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+NSS responder request setup
+---------------------------
+
+In preparation for setting up a request, The NSS responder calls into the function
+``nss_cmd_getpwnam()``. This function is a wrapper of the generic function
+``nss_getby_name()`` with the  argument ``cache_req_type`` enum ``CACHE_REQ_USER_BY_NAME``.
+The cache request types can be understood as a mapping from the equivalent NSS
+library interface lookup functions. ``getgetnam()`` sets ``CACHE_REQ_GROUP_BY_NAME``,
+``nss_cmd_getgrgid(``) sets ``CACHE_REQ_GROUP_BY_ID()``, etc...
```
Comment (elkoniu) [inline src/contrib/cache_req.rst:36]: I think I would put here full definition of `enum cache_req_type` in some scrollable box or table with some mappings in between functions and `cache_req` types.

File: src/contrib/cache_req.rst [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+to begin the cache request operation.
+
+Sending a cache request
+-----------------------
+
+``cache_req_send()`` is the top-level function to initiate a cache request, in our
+example it is called by the NSS responder ``nss_getby_name()`` function. The
+**simplified** ``cache_req_send()`` logic is described below:
+
+.. code-block:: sssd-log
```
Comment (elkoniu) [inline src/contrib/cache_req.rst:92]: This code snippet not fits the description above.

File: src/contrib/cache_req.rst [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+	[cache_req_select_domains] (0x0400): CR #10: Performing a multi-domain search
+
+.. note::
+    SSSD lookups may bypass the cache or the data provider. This determination is based on the ``cache_behavior`` enum which is also set at this time.
+    This can be seen in the logs here, `check` may be replaced with `bypass`.
+
+	*[cache_req_search_domains] (0x0400): CR #23: Search will check the cache and check the data provider*
```
Comment (elkoniu) [inline src/contrib/cache_req.rst:189]: This should be `sssd-log` snippet I think.

File: src/contrib/cache_req.rst [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+    [cache_req_set_domain] (0x0400): CR #0: Using domain [implicit_files]
+
+Performing the search
+---------------------
+
+In a multi-domain search, the below logic repeats for each domain until a result is found, or SSSD searched all
+domains and no result was found. Let's assume the normal cache behavior(do not bypass).
+
+.. mermaid::
```
Comment (elkoniu) [inline src/contrib/cache_req.rst:204]: I would use a different graph skin for the control flow and a different one for the data flow.

File: src/contrib/cache_req.rst [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    B --Not found--> D{Data Provider}
+    B --Object found--> C[Expired?]
+    C --Object expired--> D
+    D --Object found in dp--> E[Update local cache]
+    D --Object not found --> F[Next domain]
+    F --> A
+    E --> G[Search local cache again]
+
+
+* First the search checks the negative cache(``cache_req_user_by_name_ncache_check()``)
```
Comment (elkoniu) [inline src/contrib/cache_req.rst:218]: Graph mixed with bullet list and log snippets make this chapter content a bit hard to follow.

**Comment by justin-stephenson** [comment]
I did a large rewrite of the document, I believe it is more aligned with the expected content now.

I would appreciate help in completing writing about the ``Well known object`` code and proper use of the cache_req plugin well_known_fn - I did not really understand what this code does, or what problem is solves.

Thank you for the review @elkoniu but I suspect your review items may not be relevant anymore as I removed most of the previous content.

File: src/contrib/cache_req.rst [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+
+Preparing the domain
+--------------------
+
+Before performing the object search inside the domain, the ``prepare_domain_data_fn`` is called, this is used by some plugins to to alter lookup data per specific domain rules, such as case sensitivity, fully qualified format etc.
+
+Domain locator
+--------------
+
+The domain locator plugin exists to alleviate performance problems when SSSD must iterate over several domains(ex: AD forest with multiple domains) for each unqualified name, or by-ID lookups. The domain locator can help to find the correct domain to search early on in the lookup flow, instead of iterating through all domains.
```
[the remark this answers is not in the document]
Comment (pbrezina) [reply src/contrib/cache_req.rst:245]: each unqualified name, or by-ID lookups -> each unqualified name or by-ID lookups

File: src/contrib/cache_req.rst [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+
+    const struct cache_req_plugin cache_req_object_by_id = {
+        ...
+        .dp_get_domain_check_fn = cache_req_object_by_id_get_domain_check,
+        .dp_get_domain_send_fn = cache_req_object_by_id_get_domain_send,
+        .dp_get_domain_recv_fn = cache_req_common_get_acct_domain_recv,
+
+A cache-only search is performed first to check if the object already exists in the cache. If not, then we execute the domain locator plugin ``dp_get_domain_check_fn`` function to check if the id exists in the negative cache. If it is not in the negative cache, then we call the ``dp_get_domain_send_fn``
+
+The send function sends a SBUS ``getAccountDomain`` request to the backend, if a domain is reported as containing an object, all domains except that
```
Comment (pbrezina) [inline src/contrib/cache_req.rst:259]: a SBUS -> an SBUS

File: src/contrib/cache_req.rst [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+one are marked with negative cache entries for that request(using the associated ``.ncache_add_fn`` plugin function``.
```
Comment (pbrezina) [inline src/contrib/cache_req.rst:260]: plugin function``. -> plugin function.

File: src/contrib/cache_req.rst [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+
+The domain locator plugin code is only executed for unqualified requests with multiple domains, or on the second pass of a ``CACHE_REQ_CACHE_FIRST`` lookup.
```
Comment (pbrezina) [inline src/contrib/cache_req.rst:262]: domains, or on the -> domains or on the

File: src/contrib/cache_req.rst [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+
+* First the search checks the negative cache using the plugin ``ncache_check_fn`` function. If the object exists in the negative cache, and the negative cache timeout has not been reached then cache_req will return ``ENOENT``.
+
+.. code-block:: c
+
+    .ncache_check_fn = cache_req_user_by_name_ncache_check,
+
+* Next, search the cache to see if the object already exists in the local cache. Here, the plugin-defined ``lookup_fn`` is used to handle the different object types and ``SYSDB`` attributes to search. If the object is found and is not expired, it can be returned successfully.
+
+.. warning::
```
Comment (pbrezina) [inline src/contrib/cache_req.rst:296]: Place this before or after the note.

**Review by pbrezina — CHANGES_REQUESTED** [review]
cache_req -> \``cache_req``

Minor comments in review. You did a good job there.

File: src/contrib/cache_req.rst [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+If ``allow_missing_fqn`` is set to false, then this multi domain search iterating through domains will skip domains which require fully qualified names.
+
+``get_next_domain_flags`` sets the flags which are passed to ``get_next_domain()`` during iteration. At the time of this writing, cache request plugins set this to ``0``, or ``SSS_GND_DESCEND`` to determine behavior when iterating through multiple domains.
+
+.. table::
+    :align: left
+    :widths: 1, 3
+    :width: 70%
+
+    =========================== ====================================
```
Comment (pbrezina) [inline src/contrib/cache_req.rst:232]: Add a header. It will look better.

**Comment by justin-stephenson** [comment]
Thanks very much for the review, review items addressed as requested.

**Comment by pbrezina** [comment]
Ack.
